libkbfs/file_data.go

```diff
@@ -1472,47 +1704,127 @@ func (fd *fileData) deepCopy(ctx context.Context, codec kbfscodec.Codec,
 		return zeroPtr, nil, err
 	}
 
-	newTopBlock, err := topBlock.DeepCopy(codec)
-	if err != nil {
-		return zeroPtr, nil, err
-	}
-	newTopPtr = fd.rootBlockPointer()
-	if topBlock.IsInd {
-		newID, err := fd.crypto.MakeTemporaryBlockID()
+	// Handle the single-level case first.
+	if !topBlock.IsInd {
+		newTopBlock, err := topBlock.DeepCopy(codec)
 		if err != nil {
 			return zeroPtr, nil, err
 		}
-		newTopPtr = BlockPointer{
-			ID:      newID,
-			KeyGen:  fd.kmd.LatestKeyGeneration(),
-			DataVer: dataVer,
-			Context: kbfsblock.MakeFirstContext(fd.uid),
-		}
-	} else {
+
+		newTopPtr = fd.rootBlockPointer()
 		newTopPtr.RefNonce, err = fd.crypto.MakeBlockRefNonce()
 		if err != nil {
 			return zeroPtr, nil, err
 		}
 		newTopPtr.SetWriter(fd.uid)
+
+		if err = fd.cacher(newTopPtr, newTopBlock); err != nil {
+			return zeroPtr, nil, err
+		}
+
+		fd.log.CDebugf(ctx, "Deep copied file %s: %v -> %v",
+			fd.file.tailName(), fd.rootBlockPointer(), newTopPtr)
+
+		return newTopPtr, nil, nil
 	}
-	fd.log.CDebugf(ctx, "Deep copying file %s: %v -> %v",
-		fd.file.tailName(), fd.rootBlockPointer(), newTopPtr)
 
-	// Dup all of the leaf blocks.
-	// TODO: deal with multiple levels of indirection.
-	if topBlock.IsInd {
-		for i, iptr := range topBlock.IPtrs {
-			// Generate a new nonce for each one.
-			iptr.RefNonce, err = fd.crypto.MakeBlockRefNonce()
+	// For indirect files, get all the paths to leaf blocks.
```

Comment:
I think the comment is accurate -- it gets _paths_ to the leaf blocks, but not the leaf blocks themselves.  Those paths consist entirely of indirect blocks.  I guess that corresponds to what you said about the leafiest indirect blocks.

Regarding memory -- `deepCopy` is only called during conflict resolution, and then only when there's an actual conflict where two devices modified the same file.  Also, large files aren't enabled in production with this PR, they are only enabled in the next PR that also includes `IsDirect`.  So I'm not too worried about this particular function being a problem (though I suppose if you had a conflict in a GIGANTIC file, where the indirect blocks themselves could overwhelm memory, there might be an issue).  `undupChildrenInCopy` is another issue, I'll address it below.  But I'm pretty comfortable with keeping `deepCopy` as is for the near future.